src/raft/raft.rs

```diff
@@ -585,14 +648,13 @@ impl<T: Storage> Raft<T> {
     }
 
     pub fn bcast_heartbeat_with_ctx(&mut self, ctx: Option<Vec<u8>>) {
-        // TODO: avoid copy
-        let ids: Vec<_> = self.prs.keys().cloned().collect();
-        for id in ids {
-            if id == self.id {
-                continue;
+        let prs = mem::replace(&mut self.prs, ProgressSet::default());
```

Comment:
I think it's a bad idea. What if `self.prs` is queried before it's reset? It's hard to write correct code.

Reply:
How about write code like this:
```rust
struct Raft {
    prs: ProgressSet,
    raft: _Raft,  // _Raft is a struct contains all other fields
}
```
So all method of `_Raft` need pass a `&ProgressSet` or `&mut ProgressSet` in.